### DISCUSSION TITLE
docker example with php-webdriver?

### DISCUSSION THREAD

**Discussion by ildarius** [php-webdriver/php-webdriver#865] [Q&A] [answered]
Hello

I was wondering if there are any repos you can recommend which show an example of how php-webdriver can be launched in docker. 

From what I understand you'll need to instantiate 2 containers, one that starts up the java server and the other one that runs the tests?

Or does it have to be a 3 container setup such as described here? https://stackoverflow.com/questions/35175366/connection-refused-phpunit-selenium-in-docker

Any help is appreciated

**Comment by andrewnicols** [comment, marked as answer]
That's a complex question and one with many answers. Here's what I do:

I use the Selenium images available from https://hub.docker.com/u/selenium and map the port. Wherever I'm running PHP from I then just configure it to point to that port.

If I'm running PHP locally (bare metal, no docker), I am basically doing:

```
docker run --rm -d -p 4444:4444 --shm-max=2g selenium/standalone-chrome:latest
```

Then I configure php-webdriver to use a URL of `http://localhost:4444/wd/hub`

If I need to upload files, I also mount the relevant directory:
```
docker run --rm -d -p 4444:4444 -v /var/www/tests/fixtures:/var/www/tests/fixtures --shm-max=2g selenium/standalone-chrome:latest
```


If I am running from a pure-docker environment I:
```
# Create a docker network
docker network create -n testnetwork

# Create the images on that network - no need for port forwarding
docker run  --name=sel --rm -d --network testnetwork -v /var/www/tests/fixtures:/var/www/tests/fixtures --shm-max=2g selenium/standalone-chrome:latest
docker run -d --network .... my/testimage

# Configue Selenium to use a wd_host of http://sel:4444/wd/hub
```

Note: our names are dynamically created based on the jobid and configuration.

Our CI scripts are available at https://github.com/moodlehq/moodle-ci-runner/blob/master/runner/master/run.sh if you're interested.

**Reply by ildarius** [reply]
Thank you so much! Just to confirm, once you startup the container

`docker run --rm -d -p 4444:4444 --shm-max=2g selenium/standalone-chrome:latest
`

Do you then have another container that runs your tests and that points to that URL?

Also just to confirm. Your PHP tests that point to that URL. It still needs to have the Chromedriver + Chrome installed on it right? Or that part is taken care of in the containers that spawns the server?

**Reply by andrewnicols** [reply]
If I'm running the tests natively - i.e. on my local machine, then I don't have another container for running the tests. I just use `selenium/standalone-chrome:latest` to provide selenium. chrome. chromedriver. PHP is installed on my laptop and the tests are run _from_ my laptop.

I _can_ have a separate container to run the tests _from_ - that's where I do that extra step to create a docker network, and then put the selenium and PHP image into that container. SO `my/testimage` may be something like `php/7.4-apache`.

The selenium container runs Selenium, and includes chrome and chromedriver. The `selenium/standalone-chrome` image is a single image which includes Chrome, Chromedriver, and a standalone variant of Selenium. You can use a grid version, and then register the individual browsers to it, but I wouldn't recommend it unless you need to.

**Comment by ildarius** [comment]
EDITED

Following reply from andrew should be used for a pure docker setup (1 container with chrome, chromedriver and 1 container with the actual tests). 

```
# Create a docker network
docker network create -n testnetwork

# Create the images on that network - no need for port forwarding
docker run  --name=sel --rm -d --network testnetwork -v /var/www/tests/fixtures:/var/www/tests/fixtures --shm-max=2g selenium/standalone-chrome:latest
docker run -d --network .... my/testimage

# Configue Selenium to use a wd_host of http://sel:4444/wd/hub
```

**Reply by andrewnicols** [reply]
Thanks for that write-up.

In your step 2, I don't think that you need to add selenium or Chrome - they are provided by the selenium-hub host.

I'll look at turning the above into an example when (if) I have some free time ...

**Reply by ildarius** [reply]
Here's a super simple repository which does what you described in your answer, but in a slightly different way

https://github.com/ildarius/docker_phpwebdriver
